CMakeLists.txt

```diff
@@ -720,34 +723,20 @@ if(WITH_OPTIM)
         if(MSVC)
             list(APPEND ZLIB_ARCH_HDRS fallback_builtins.h)
         endif()
-        if(WITH_AVX2 AND HAVE_AVX2_INTRIN)
-            add_definitions(-DX86_AVX2 -DX86_AVX2_ADLER32)
-            list(APPEND ZLIB_ARCH_SRCS ${ARCHDIR}/slide_avx.c ${ARCHDIR}/adler32_avx.c)
-            add_feature_info(AVX2_SLIDEHASH 1 "Support AVX2 optimized slide_hash, using \"${AVX2FLAG}\"")
-            list(APPEND ZLIB_ARCH_SRCS ${ARCHDIR}/compare258_avx.c)
-            add_feature_info(AVX2_COMPARE258 1 "Support AVX2 optimized compare258, using \"${AVX2FLAG}\"")
-            add_feature_info(AVX2_ADLER32 1 "Support AVX2-accelerated adler32, using \"${AVX2FLAG}\"")
-            add_intrinsics_option("${AVX2FLAG}")
-        endif()
         if(WITH_SSE4 AND (HAVE_SSE42CRC_INLINE_ASM OR HAVE_SSE42CRC_INTRIN))
             add_definitions(-DX86_SSE42_CRC_HASH)
             list(APPEND ZLIB_ARCH_SRCS ${ARCHDIR}/insert_string_sse.c)
+            add_intrinsics_option("${SSE4FLAG}" ${ARCHDIR}/insert_string_sse.c)
             add_feature_info(SSE42_CRC 1 "Support SSE4.2 optimized CRC hash generation, using \"${SSE4FLAG}\"")
-            add_intrinsics_option("${SSE4FLAG}")
             if(HAVE_SSE42CRC_INTRIN)
                 add_definitions(-DX86_SSE42_CRC_INTRIN)
             endif()
         endif()
-        if(HAVE_SSE42CMPSTR_INTRIN)
-            add_definitions(-DX86_SSE42_CMP_STR)
-            list(APPEND ZLIB_ARCH_SRCS ${ARCHDIR}/compare258_sse.c)
-            add_feature_info(SSE42_COMPARE258 1 "Support SSE4.2 optimized compare258, using \"${SSE4FLAG}\"")
-        endif()
         if(WITH_SSE2 AND HAVE_SSE2_INTRIN)
             add_definitions(-DX86_SSE2)
             list(APPEND ZLIB_ARCH_SRCS ${ARCHDIR}/slide_sse.c)
             if(NOT ${ARCH} MATCHES "x86_64")
-                add_intrinsics_option("${SSE2FLAG}")
+                add_intrinsics_option("${SSE2FLAG}" ${ARCHDIR}/slide_sse.c)
                 add_feature_info(FORCE_SSE2 FORCE_SSE2 "Assume CPU is SSE2 capable")
```

Comment:
Just some thoughts while reviewing this:

FORCE_SSE2 should also apply SSE2 to every compiled file, meaning we require the cpu to be SSE2 capable and that the compiler is free to utilize SSE2 and any preceding instruction set anywhere in the code. (This is only a base instruction set, we will still add AVX2 for files that need that for example).

WITH_SSE2 should not however, that should only apply to slide_sse.c for example.

I don't think FORCE_SSE2 should depend on WITH_SSE2 though, since the latter enables SSE2-specific functions, and the former enables the instruction set (but the X85_NOCHECK_SSE2 define still needs to depend on both being active though).

Looks like this distinction has been lost at some point (or I planned for this and forgot about it long ago, not sure).
I don't think FORCE_SSE2 currently exists in configure, but I think that would be good thing to add there without it requiring a lot of code.

I also think FORCE_SSE2 should default to ON on x86 (and permanently ON on x86_64), and that anyone who wants to support older cpus will be required to disable it manually. Those cpus are quite old by now, default to ON and get the speedups for free by default.

This might be a good fit for a separate PR though, but I will accept it as a separate commit in this PR if you want to implement it.